### PR TITLE
[KED-2163] Add architecture doc

### PR DESCRIPTION
**Pull request by studioswong**
## Description

Related tickets: [KED-2467](https://jira.quantumblack.com/browse/KED-2467), [KED-2468](https://jira.quantumblack.com/browse/KED-2468) and  [KED-2467](https://jira.quantumblack.com/browse/KED-2467)

This is part 1 of setting up the Architecture docs which includes the high-level overview as well as the codemap. Part 2 will include the documentation of cross-cutting concerns, as well as supplementary diagrams to illustrate the overall structure , actions, and data flow of the app. 

Please note this is the very first draft and an initial set up of the docs to initially gauge your feedback on the topics / areas of the app / how detailed I would go, etc. I would partcularly like your feedback on whether the document is clear enough to outline our app for someone new to the project, as well as whether I had left out any important concepts that should be mentioned here. 

And of course, this doc will continue to be updated as our application evolve over time, hence we can also aways revisit this doc. 

Target audience for these docs:

 * new joiners into the Kedro-viz project
 * open source developers hoping to contribute to the project
 * anyone who is interested in how kedro-viz works under the hood 

## Development notes

I have set up a `ARCHITECTURE.md` file under the main directory. ( The idea of setting up a directory for the docs was raised on slack, but given the current amount of docs there is not a need to do so yet. We can revisit this idea if our docs continue to expand. 

There are also 3 diagrams that accompanies this doc. 

## QA notes

N/A

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

## Legal notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: ARCHITECTURE.md
Comment (richardwestenra) [inline ARCHITECTURE.md:44]: We want to keep this as brief and high-level as possible - we should avoid describing how each file works - you can use in-file comments for this!

Instead, we want a brief, high-level overview of the data flow, specifically focused on why we do things the way we do. e.g. something like:

> On initialisation, the app uses a string token (e.g. 'json' or 'animals' to determine the data source - you can find the details about which identifier refers to which data source in [CONTRIBUTING.md](https://github.com/quantumblacklabs/kedro-viz/blob/main/CONTRIBUTING.md#data-sources). The application then [manually normalises this data](https://github.com/quantumblacklabs/kedro-viz/blob/main/src/store/normalize-data.js), in order to [make immutable state updates as performant as possible](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape).
> Next, the app [initialises the Redux data store](https://github.com/quantumblacklabs/kedro-viz/blob/main/src/store/initial-state.js), by merging together initial defaults, `window.localStorage` data, and the normalized pipeline data. The initial state data is separated into two: Pipeline and non-pipeline state. This is because the non-pipeline state should persist if the pipeline data is reset - e.g. if the user selects a new top-level pipeline.

or

> When loading asynchronous data (specified using the `json` identifier), the app first loads the `/api/main` endpoint. This is because it contains the list of available top-level pipelines, which is necessary to check before loading others. If another pipeline's data is required (which happens under certain conditions, e.g. [...]), it then loads that pipeline's data from `/api/pipeline/[ID]` and replaces the app state with this new data. Conversely, when the user selects a node on the graph, if this node has not previously been selected then the app will request data for this node from `/api/nodes/[ID]`.

You could write this a lot more accurately and informatively, this is just a couple of examples of the top of my head to give an indication of the level of depth that we should aim for. Focus on _what_ and _why_, not _how_. If you are unsure _why_ we do something the way we do, please ask! :)

**Review by richardwestenra — CHANGES_REQUESTED** [review]
Thanks Susanna! This is a great start, and you've given me a lot to think about. Let's pair on this for the next phase?

File: ARCHITECTURE.md
Comment (richardwestenra) [reply ARCHITECTURE.md:44]: I've had another look at [an example architecture doc](https://github.com/rust-analyzer/rust-analyzer/blob/d7c99931d05e3723d878bea5dc26766791fa4e69/docs/dev/architecture.md) and it does have specific description of parts of the codebase, but more describing sections/directories rather than individual files. Maybe we should approach it like that. We should decide what the best taxonomy/breakdown would be, and how granular we need to get. Ideally as little granularity as possible, while still being descriptive.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:12]: Thanks @richardwestenra ! Yes I completely agree on further elaborating on that difference and explain the detailed concepts of `nodes` and `edges` in kedro-viz. 

Speaking of, what exactly is defined as a concept of an `edge` in Kedro-viz?

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:14]: good point - I will state that about the API and add a section about the back-end.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:16]: good point! I agree on the order in which the endpoints should be listed is key. 

Re: the conditions which they are called - I thought about elaborating on that while writing this document, but worry this will go into the territory of too much details, and moreover given that the conditions to call those endpoints could change drastically depending on future work, which would lead to the need to constantly revise this document, hence I was thinking we could keep it as general as possible to only state the avaliable endpoints. What do you think? We could further discuss on that when we pair.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:18]: another great point about how those 2 cases affects data loading - I will further elaborate on that 👍 

I will modify and further elaborate about how data is consumed via a JS object prop for the react component as well.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:24]: oh indeed - I keep thinking we're using the redux-toolkit API ( such as `configureStore`) but actually it's our own function 👍 

yea I agree this section should be renamed as `State management and data flow`

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:39]: I totally agree on removing the tools section - these could be covered in the style guide for sure.

I'll instead mention the tools we use within the codemap section under comments on the relating file / function

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:46]: good point - I'll add in the full path

File: ARCHITECTURE.md
Comment (richardwestenra) [reply ARCHITECTURE.md:12]: An edge is a link between any two nodes (i.e. nodes, datasets or parameters). It's represented by a line on the graph.

File: ARCHITECTURE.md
Comment (richardwestenra) [reply ARCHITECTURE.md:16]: I think that the process by which data is requested and loaded into the state is probably the most important thing we could cover in this document. I don't think it's so brittle that it would change in a hurry.

Also fwiw, there are more than 2 endpoints: There's `/main` as well. Also they all begin with `/api`. See the `getUrl` utility.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:18]: @richardwestenra I will add a new section on data, which includes details on data loading, along with the data-flowdiagram.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:44]: Thanks @richardwestenra , that's a great doc ( which the same doc has been heavily referenced for the structure and format of this doc too :D )

I have referenced your suggested paragraph in the newly added dataflow section (which also includes the data-flow diagram that we flushed out). Will let you know when I have pushed it and let me know your thoughts!

Comment (studioswong) [reply ARCHITECTURE.md:55]: I'm added a new data flow section that incorporates the ideas you suggested above.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:23]: This is now stated in the newly added data section

File: ARCHITECTURE.md
Comment (studioswong) [inline ARCHITECTURE.md:115]: @bru5 , let me know if there is anything you think I should elaborate  / expand on for this section. I kept it as simple as possible to avoid the need to constantly main this doc - the new architecture diagram with the actions for the minimap and flowchart should imply a lot of ideas under the hood.

**Comment by studioswong** [comment]
currently some related python tests are failing - see below. 
image

**Comment by stichbury** [comment]
I've taken a look at this page -- is it ready for me to review/edit or do you want to resolve the outstanding comments and let me know when to go in and make final adjustments.

My main comment is to meet the style guide for Kedro docs, which is to use UK English and simplify as much as possible (don't use "utilize" if you can use "use" for example). There are a few typos too. I don't mind fixing all this, but would rather do it at the end of the chain than when you're still working on the page. Just LMK when you're ready!

**Comment by richardwestenra** [comment]
@stichbury Thanks Jo - I'm doing a fairly extensive edit/rewrite so I'll ping you when it's ready for you to take a look

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md]: resolving this since this section has been rewritten and this comment is not relevant anymore.

**Comment by stichbury** [comment]
> @stichbury Hey Jo, I've finished my edits/rewrites, it's ready for you to take a look! Thanks :)

Thanks @richardwestenra -- I shall do the review tomorrow (Thursday 11th)

**Review by stichbury — APPROVED** [review]
Awesome work. Thanks for the preview and apologies again for the delay in getting this complete. 🥇

File: ARCHITECTURE.md
Comment (yetudada) [inline ARCHITECTURE.md:49]:
```suggestion
Each pipeline endpoint corresponds to a different registered pipeline. Only one registered pipeline should be loaded at a time, so loading data from a pipeline endpoint will reset the pipeline state in the store. Each pipeline dataset contains all the data required to render the graph.
```

Comment (yetudada) [reply ARCHITECTURE.md:49]: A top-level pipeline is actually a _registered pipeline_. That's what we make users do, they register their pipelines.

File: ARCHITECTURE.md
Comment (yetudada) [inline ARCHITECTURE.md:115]:
```suggestion
- `task`: a Kedro [node](https://kedro.readthedocs.io/en/stable/13_resources/02_glossary.html#node), i.e. a Python function wrapper
```

**Review by yetudada — APPROVED** [review]
Such great documentation! I've made minor comments.

**Review by richardwestenra — APPROVED** [review]

File: ARCHITECTURE.md
Comment (bru5) [inline ARCHITECTURE.md:49]: Should possibly clarify "registered pipeline" in the first sentence here e.g. "registered pipeline in the Kedro project" as a new contributor won't be familiar with that concept yet I think.

File: ARCHITECTURE.md
Comment (bru5) [inline ARCHITECTURE.md:51]: Similarly could clarify "active pipeline" as "user's active pipeline"? Could the last sentence be split in two to make it a bit simpler?

File: ARCHITECTURE.md
Comment (richardwestenra) [reply ARCHITECTURE.md:12]: I forgot to add anything about the layout engine, maybe we should add a new section about it as part of a subsequent PR?

**Comment by richardwestenra** [comment]
@bru5 I've updated the wording based on your suggestions - thanks! I'd suggest that we merge this now as it's got 3 approvals, and if there are any further changes/additions to be made, I think we can make them as part of a subsequent PR.

File: ARCHITECTURE.md
Comment (studioswong) [reply ARCHITECTURE.md:12]: yea I agree, or even the layout engine and its logic behind can be a seperate doc on its own?

**Comment by noklam** [comment]
@studioswong Randomly bump into this document, it's awesome and well written!

**Pull request merged**
